Common error details, as @newpavlov's design

dhardy opened the pull request:
Implement new design in #10.

Any comments @newpavlov or @pitdicker?

`Error` should be 32 bytes in size (two enums padded for alignment + a fat pointer). It could potentially be reduced to 24 bytes by combining the two enums, but that would be ugly. Otherwise it could only be much smaller by omitting the details altogether. Hopefully the size doesn't make too much of a difference anyway; I don't know.

Most of the "details" implementation is hidden and "kind" is still not exhaustively matchable.

[Doc here](https://dhardy.github.io/rand/rand_core/struct.Error.html)

pitdicker commented on rand_core/src/lib.rs:
```diff
@@ -247,55 +247,91 @@ pub enum ErrorKind {
-        }
+        write!(f, "RNG {} (details: {})",
+            self.kind.description(),
+            self.details().unwrap_or("NA"))
```
We maybe should not print the details here, see https://github.com/rust-lang-nursery/error-chain/issues/154#issuecomment-305936397

pitdicker commented on rand_core/src/lib.rs:
```diff
@@ -247,55 +247,91 @@ pub enum ErrorKind {
-        self.cause.as_ref().map(|e| &**e)
+        match self.details {
+            ErrorDetails::Error(ref e) => Some(&**e),
+            // unforunately we cannot report &str cause here
```
I think we can if `ErrorDetails` where to implement `std::error::Error`. Or by implementing `std::error::Error` for `ErrorDetails::Str(...)`

pitdicker commented on rand_core/src/lib.rs:
```diff
@@ -247,55 +247,91 @@ pub enum ErrorKind {
+    /// Create a new instance, with a static string describing the details.
+    /// 
+    /// This may be used in both `std` and `no_std` enviroments.
+    pub fn new_str(kind: ErrorKind, details: &'static str) -> Self {
```
The name could be a little bit confusing. Maybe a more verbose `new_with_cause_str`?

dhardy replied:
Why the long name?

newpavlov replied:
I agree with `new_str` being a bit confusing, but I personally can't come up with the better alternative, so I guess we can leave it up to docs.

pitdicker commented on rand_core/src/lib.rs:
```diff
@@ -247,55 +247,91 @@ pub enum ErrorKind {
-        Error {
-            kind: kind,
-            cause: cause,
+    pub fn new_err<E>(kind: ErrorKind, cause: E) -> Self
```
Does it make sense to have a function that takes both a kind and a string, and use one of them depending on whether the `std` feature is enabled? Now implementations of `Rng` need a `std` feature flag to give the right type of cause.

dhardy replied:
I don't get what you mean? This is more generic than `new_str` and can still accept a string but is only available with the `std` library (we could probably depend on `alloc` instead, but that's another matter; we need `Box`).

dhardy replied:
See [this comment](https://github.com/dhardy/rand/issues/10#issuecomment-338666735).

pitdicker replied:
The `Into` where clause it neat, wouldn't have thought about that myself.  Although maybe it is common, and I am just missing something :-).

I was thinking about cases like [this](https://github.com/dhardy/rand/pull/28/files#diff-412db97196316ae7812d61b5624f0ad9R162), where a `Rng` implementation needs to use different functions to construct an error depending on whether the `std` feature is available.

Then a helper function is maybe useful:
``` rust
The `Into` where clause it neat, wouldn't have thought about that myself.  Although maybe it is common, and I am just missing something :-).

``` rust
#[cfg(feature="std")]
pub fn new_with_cause(kind: ErrorKind, cause: E, details: &'static str) -> Self {
    new_str(kind, details)
}

#[cfg(not(feature="std"))]
pub fn new_with_cause(kind: ErrorKind, cause: E, details: &'static str) -> Self {
    new_err(kind, cause)
}
```

newpavlov replied:
In `no_std` environment you simply can't have implementation of `Error` trait, as it's defined only in the `std`. So either way crate authors will have to write `cfg`s to handle switching.

dhardy replied:
Or authors supporting `no_std` can simply use `new_str` with a string description everywhere. There's no advantage to using `new_err` if the details are just a string.

pitdicker replied:
I updated my branch with `JitterRng` to the new error handling https://github.com/pitdicker/rand/commit/b223c60539680490c98316fd1d4f8b3dcb8b17ca.

The second commit shows what I mean https://github.com/pitdicker/rand/commit/f6b353154b8dc8083ddb78f77ad1f91634098b8e.
In theory an RNG implementation can support both `std` and `no_std` compatible error causes, and the choice is made in `rand_core`.

pitdicker replied:
It was just an idea, but I don't think it helps in enough cases. Lets leave the two functions as is.

dhardy replied:
Your `new_with_cause` isn't even going to compile in `no_std` mode since it depends on `std::error::Error`.

I see what you mean, but I'm not convinced it's worth returning a chained `Error` type in this case; there is no added information (but maybe your `description()` strings should include all the details from the `Display` impl).

pitdicker replied:
Oops. Ok, please forget I said anything :-)

newpavlov commented on rand_core/src/lib.rs:
```diff
@@ -247,55 +247,138 @@ pub enum ErrorKind {
+    /// 
+    /// In the case of a chained error, the actual error may be obtained via
+    /// `std::error::Error::cause`.
+    pub fn details(&self) -> Option<&str> {
```
Maybe "will be obtained"? Also not `description`, but `cause`.

dhardy replied:
Um, actually I think my description is correct, but I'll try to make it clearer.

newpavlov replied:
I meant in reverse. :) In the code `description()` method is used, but in the docstring you write about `cause`.

dhardy replied:
This is the `details()` function. Anyway, the terms *are* a bit confused, although `description` should be distinct from `details`/`cause`

newpavlov commented on rand_core/src/lib.rs:
```diff
@@ -247,55 +247,138 @@ pub enum ErrorKind {
-            ErrorKind::Other => write!(f, "An unspecified RNG error occurred."),
-            ErrorKind::__Nonexhaustive => unreachable!(),
-        }
+        write!(f, "RNG {}", self.kind.description())
```
I think "RNG Error: {}" will be better.

newpavlov commented on rand_core/src/lib.rs:
```diff
@@ -247,55 +247,138 @@ pub enum ErrorKind {
+impl ::std::error::Error for ErrorDetails {
+    fn description(&self) -> &str {
+        match *self {
+            ErrorDetails::None => "NA", // case shouldn't be used
```
It's better to write `unreachable!()` here if this case indeed can not be used anywhere.

dhardy commented:
Ok, I addressed some of your points.

I don't see much point renaming the `new_*` functions. As it is they are descriptive of the *type* of the details (`str` or something coercible to an `Error`). We could renamed to `new_with_*`, but that's longer for little extra clarity. We could use names like `new_cause` (or `new_chained`) and `new_msg` instead, but I'm not sure that's any clearer.

pitdicker commented on rand_core/src/lib.rs:
```diff
@@ -247,55 +247,136 @@ pub enum ErrorKind {
-            ErrorKind::Other => write!(f, "An unspecified RNG error occurred."),
-            ErrorKind::__Nonexhaustive => unreachable!(),
-        }
+        write!(f, "RNG error: {}", self.kind.description())
```
According to the [book](https://doc.rust-lang.org/book/first-edition/error-handling.html#the-error-trait) `description` is supposed to be a short description, and `Display` a user-facing representation of the error (a proper sentence with capital letter, dot at the end, and a bit less condensed text).

I think this is okay, but like the previous version more.

dhardy replied:
Does it? I skimmed through and saw an example similar to this, but in general the section is long and a little vague.
